script/model/model_server.py

```diff
@@ -0,0 +1,355 @@
+        :param send_id: id on this end, 0 for now
+        :param recv_id: callback id to invoke on the other end
+        :param data: payload of the message in JSON
+        :return:
+        """
+        json_result = json.dumps(data)
+        msg = f"{send_id}-{recv_id}-{json_result}"
+        self.socket.send_multipart([''.encode('utf-8'), msg.encode('utf-8')])
+
+    @staticmethod
+    def _make_response(action: Callback, result: Any) -> Dict:
+        """
+        Construct a response to the ModelServerManager
+        :param action:  Action callback on the ModelServerManager
+        :param result: Any result
+        :return:
+        """
+        return {
+            "action": action,
+            "result": result
+        }
+
+    @staticmethod
+    def _parse_msg(payload:str) -> Tuple[int, int, Message]:
+        logging.debug("PY RECV: " + payload)
+        tokens = payload.split('-', 2)
+
+        # Invalid message format
+        try:
+            msg_id = int(tokens[0])
+            recv_id = int(tokens[1])
+        except ValueError as e:
+            logging.error(f"Invalid message payload format: {payload}, ids not int.")
+            return -1, -1, None
+
+        msg = Message.from_json(tokens[2])
+        return msg_id, recv_id, msg
+
+    @staticmethod
+    def _transform_opunit_data(raw_data: Dict) -> List[opunit_data.OpUnitData]:
+        """
+        Transform the raw training data from ModelServerManger for each data unit
+        :param raw_data: {
+            opunit: (x, y)
+        }
+        :return:
+        """
+        # TODO(ricky): serialize the opunit by name rather than index
+        return [opunit_data.OpUnitData(OpUnit(x[0]), x[1][0], x[1][1]) for x in raw_data]
+
+    def _train_model(self, data: Dict) -> str:
+        """
+        Train a model with the given model name and seq_files directory
+        :param data: {
+            models: [lr, XXX, ...],
+            seq_files: PATH_TO_SEQ_FILES_FOLDER, or None
+            data_lists: [(OpUnit, X_List, Y_List)/OpUnitData]
+        }
+        :return: the model map
+        """
+        ml_models = data["models"]
+        seq_files_dir = data["seq_files"]
+        raw_data = data["raw_data"]
+
+        result_path = ModelServer.RESULT_PATH
+        test_ratio = ModelServer.TEST_RATIO
+        trim = ModelServer.TRIM_RATIO
+        expose_all = ModelServer.EXPOSE_ALL
+        txn_sample_interval = ModelServer.TXN_SAMPLE_INTERVAL
+        trainer = MiniTrainer(seq_files_dir, result_path, ml_models, test_ratio, trim, expose_all,txn_sample_interval)
+        if seq_files_dir:
+            # Perform training from MiniTrainer and input files directory
+            model_map = trainer.train()
+        else:
+            """
+            NOTE:
+            This branch has not been tested yet because the MiniTrainer now heavily relies on the CSV file. 
+            So training with data directly is not yet possible without refactoring the MiniTrainer 
```

Comment:
I don't really think we need this branch... when do you think we need to train with data directly? Where this "direct" data is going to come from?

Reply:
Yeah, I thought we could get rid of the `mini_runners` dependency by just feeding the model server with structurally identical fake data, because I feel having a test that depends on a benchmark might involve too much coupling. 
But I think feeding fake direct data to the model server couldn't be done without much refactoring, so I did not go down this path.

Reply:
Right I think we can remove this branch then.